network/vif-route-qubes

```diff
@@ -112,38 +112,58 @@ case "$command" in
         ;;
 esac
 
-domid=${vif/vif/}
-domid=${domid/.*/}
+[[ $vif =~ ^vif([0-9]+)\.([0-9]+)$ ]] || exit 1
+domid=${BASH_REMATCH[1]} sub=${BASH_REMATCH[2]}
 # metric must be possitive, but prefer later interface
 #  32752 is max XID aka domid
 metric=$(( 32752 - domid ))
 
-
+# add anti-spoofing rules before enabling the interface
 if [ "${ip}" ]; then
     # If we've been given a list of IP addresses, then add routes from dom0 to
     # the guest using those addresses.
     for addr in ${ip};
     do
-        if [[ "$addr" = *:* ]] && [[ "$ipv6_disabled" = 1 ]]; then
-            log error "Cannot set IPv6 route to ${addr}, IPv6 disabled in the kernel"
-            continue
-        fi
-        ${cmdprefix} ip route "${ipcmd}" "${addr}" dev "${vif}" metric "$metric"
         if [[ "$addr" = *:* ]]; then
             ipt=ip6tables-restore
         else
             ipt=iptables-restore
         fi
-        echo -e "*raw\\n$iptables_cmd -i ${vif} ! -s ${addr} -j DROP\\nCOMMIT" | \
+        printf '%s\n' "*raw" \
+            "$iptables_cmd -i ${vif} ! -s ${addr} -j DROP" \
+            "$iptables_cmd ! -i vif+ -s ${addr} -j DROP" \
+            "COMMIT" | \
             ${cmdprefix} $ipt --noflush $ipt_arg
-
-        network_hooks "${command}" "${vif}" "${addr}"
+        if [[ "$command" = 'online' ]]; then
+            mac=$(xenstore-read "backend/vif/$domid/$sub/mac") &&
+            ip -- neighbour "${ipcmd}" to "${addr}" \
+                dev "${vif}" lladdr "$mac" nud permanent
+        fi
     done
     # if no IPv6 is assigned, block all IPv6 traffic on that interface
     if ! [[ "$ip" = *:* ]]; then
         echo -e "*raw\\n$iptables_cmd -i ${vif} -j DROP\\nCOMMIT" | \
             ${cmdprefix} ip6tables-restore --noflush $ipt_arg
     fi
+fi
+
+if [ "$command" = "online" ]; then
+    ifconfig "${vif}" up
+fi
+
+if [ "${ip}" ]; then
+    # If we've been given a list of IP addresses, then add routes from dom0 to
+    # the guest using those addresses.
+    for addr in ${ip};
+    do
+        if [[ "$addr" = *:* ]] && [[ "$ipv6_disabled" = 1 ]]; then
+            log error "Cannot set IPv6 route to ${addr}, IPv6 disabled in the kernel"
+            continue
+        fi
+        ${cmdprefix} ip route "${ipcmd}" "${addr}" dev "${vif}" metric "$metric" onlink
```

Comment:
With this PR applied, network doesn't work at all. I've got `Error: Invalid flags for nexthop - PERVASIVE and ONLINK can not be set.` error in `/var/log/xen/xen-hotplug.log` in sys-net and I think it comes from this line.

Reply:
Indeed it does.  Good catch!